File: ios/FindMyBay/Features/Booking/SlotPickerViewModel.swift
// SlotPickerViewModel.swift
//
// Drives the slot picker — service selector, date strip (today + next 3 days),
// and a list of bookable time slots. On confirm, posts to `/bookings` and
// surfaces the new Booking so the UI can navigate to ReviewPay.
//
// Mirrors `feature/booking/SlotPickerViewModel.kt`.

import Foundation
import Observation

@Observable @MainActor
final class SlotPickerViewModel {

    // MARK: - State
    var loading: Bool = false
    var vendor: VendorDetail? = nil
    var selectedService: Service? = nil
    var date: Date = Date()
    var slots: [Slot] = []
    var selectedSlot: Slot? = nil
    var booking: Booking? = nil
    var confirming: Bool = false
    var error: String? = nil

    /// Promo accordion state — see PromoCodeBlock in SlotPickerView. Sent
    /// to the backend on confirm. Validation happens server-side; we don't
    /// pre-check the code here.
    var promoCode: String = ""
    var promoExpanded: Bool = false

    /// Today + next 3 days, normalised to start-of-day.
    var availableDates: [Date] {
        let cal = Calendar.current
        let start = cal.startOfDay(for: Date())
        return (0..<4).compactMap { cal.date(byAdding: .day, value: $0, to: start) }
    }

    // MARK: - Deps
    let vendorId: String
    /// Service id chosen on Vendor Detail. We match against the loaded
    /// vendor's services list and fall back to `.first` if missing — covers
    /// deep-links that opened the slot picker without an explicit service.
    private let initialServiceId: String?
    private let vendors: VendorRepository
    private let bookings: BookingRepository

    init(vendorId: String,
         initialServiceId: String? = nil,
         vendors: VendorRepository,
         bookings: BookingRepository) {
        self.vendorId = vendorId
        self.initialServiceId = initialServiceId
        self.vendors = vendors
        self.bookings = bookings
    }

    // MARK: - Loaders

    func loadVendor() {
        loading = true
        error = nil
        Task {
            do {
                let v = try await vendors.detail(id: vendorId)
                vendor = v
                // Honour the service the user picked on Vendor Detail. If
                // the id doesn't match any returned service (stale link,
                // deleted service), fall back to the first one so the
                // picker still shows something usable.
                if selectedService == nil {
                    selectedService = v.services.first { $0.id == initialServiceId }
                        ?? v.services.first
                }
                loading = false
                if selectedService != nil { loadSlots() }
            } catch {
                loading = false
                self.error = (error as? LocalizedError)?.errorDescription
                    ?? "Couldn't load vendor"
            }
        }
    }

    func selectService(_ s: Service) {
        selectedService = s
        selectedSlot = nil
        loadSlots()
    }

    func selectDate(_ newDate: Date) {
        if Calendar.current.isDate(newDate, inSameDayAs: date) { return }
        date = newDate
        selectedSlot = nil
        loadSlots()
    }

    private func loadSlots() {
        guard let service = selectedService else { return }
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = TimeZone.current
        let dateStr = formatter.string(from: date)
        loading = true
        error = nil
        slots = []
        Task {
            do {
                let resp = try await vendors.availability(
                    vendorId: vendorId, serviceId: service.id, date: dateStr
                )
                slots = resp.slots
                loading = false
            } catch {
                loading = false
                self.error = (error as? LocalizedError)?.errorDescription
                    ?? "Couldn't load slots"
            }
        }
    }

    func selectSlot(_ slot: Slot) {
        if slot.available { selectedSlot = slot }
    }

    // MARK: - Confirm

    /// Creates the booking and stores the result on `booking`. Caller observes
    /// that property to navigate to ReviewPay. Promo errors (`promo_*` codes
    /// from the backend, all 409 except `promo_not_found` which is 404) are
    /// surfaced verbatim via `APIError.userMessage` so the user sees the
    /// real reason ("Promo can't be applied: not started", etc.) rather than
    /// a generic "Couldn't confirm booking".
    func confirm() {
        guard let service = selectedService, let slot = selectedSlot else { return }
        confirming = true
        error = nil
        let codeToSend = promoCode
            .trimmingCharacters(in: .whitespaces)
            .uppercased()
        Task {
            do {
                let b = try await bookings.create(
                    vendorId: vendorId,
                    serviceId: service.id,
                    slotStartIso: slot.startsAt,
                    promoCode: codeToSend.isEmpty ? nil : codeToSend
                )
                booking = b
                confirming = false
            } catch let api as APIError {
                confirming = false
                self.error = api.userMessage
            } catch {
                confirming = false
                self.error = (error as? LocalizedError)?.errorDescription
                    ?? "Couldn't confirm booking"
            }
        }
    }
}

File: android/app/src/main/java/ae/findmybay/feature/booking/SlotPickerViewModel.kt
package ae.findmybay.feature.booking

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import ae.findmybay.data.repo.BookingRepository
import ae.findmybay.data.repo.VendorRepository
import ae.findmybay.domain.model.Booking
import ae.findmybay.domain.model.Service
import ae.findmybay.domain.model.Slot
import ae.findmybay.domain.model.VendorDetail
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import javax.inject.Inject

data class SlotPickerUiState(
    val loading: Boolean = false,
    val vendor: VendorDetail? = null,
    val selectedService: Service? = null,
    val date: LocalDate = LocalDate.now(),
    val availableDates: List<LocalDate> = (0..3).map { LocalDate.now().plusDays(it.toLong()) },
    val slots: List<Slot> = emptyList(),
    val selectedSlot: Slot? = null,
    val booking: Booking? = null,
    val confirming: Boolean = false,
    val error: String? = null,
    // Promo-code state. The user types into `promoCodeInput` and hits the
    // confirm button — the code is sent server-side which is the source
    // of truth for whether it's valid. On success the booking DTO comes
    // back with discountAed > 0 + promoCode populated; on failure we get
    // a 4xx and surface its message via `error`.
    val promoCodeInput: String = "",
    val promoExpanded: Boolean = false,
)

@HiltViewModel
class SlotPickerViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val vendors: VendorRepository,
    private val bookings: BookingRepository,
) : ViewModel() {

    val vendorId: String = checkNotNull(savedStateHandle["vendorId"])
    // Optional initial service id, forwarded from the vendor detail screen
    // when the customer picked a non-default service before tapping Book.
    // Null when arriving via a deep link or back-stack restore.
    private val initialServiceId: String? = savedStateHandle["serviceId"]

    private val _state = MutableStateFlow(SlotPickerUiState())
    val state: StateFlow<SlotPickerUiState> = _state.asStateFlow()

    init { loadVendor() }

    private fun loadVendor() {
        _state.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching { vendors.detail(vendorId) }
                .onSuccess { v ->
                    // Honour the forwarded serviceId if it matches one of the
                    // vendor's services; otherwise default to the first
                    // (legacy behaviour for deep links / restored state).
                    val preselected = initialServiceId?.let { id ->
                        v.services.firstOrNull { it.id == id }
                    } ?: v.services.firstOrNull()
                    _state.update { it.copy(loading = false, vendor = v, selectedService = preselected) }
                    if (preselected != null) loadSlots()
                }
                .onFailure { e -> _state.update { it.copy(loading = false, error = e.message ?: "Couldn't load vendor") } }
        }
    }

    fun selectService(service: Service) {
        _state.update { it.copy(selectedService = service, selectedSlot = null) }
        loadSlots()
    }

    fun selectDate(date: LocalDate) {
        if (date == _state.value.date) return
        _state.update { it.copy(date = date, selectedSlot = null) }
        loadSlots()
    }

    private fun loadSlots() {
        val service = _state.value.selectedService ?: return
        val date = _state.value.date.format(DateTimeFormatter.ISO_LOCAL_DATE)
        _state.update { it.copy(loading = true, error = null, slots = emptyList()) }
        viewModelScope.launch {
            runCatching { vendors.availability(vendorId, service.id, date) }
                .onSuccess { resp -> _state.update { it.copy(loading = false, slots = resp.slots) } }
                .onFailure { e -> _state.update { it.copy(loading = false, error = e.message ?: "Couldn't load slots") } }
        }
    }

    fun selectSlot(slot: Slot) {
        if (slot.available) _state.update { it.copy(selectedSlot = slot) }
    }

    fun setPromoCode(value: String) {
        _state.update { it.copy(promoCodeInput = value, error = null) }
    }

    fun togglePromoExpanded() {
        _state.update { it.copy(promoExpanded = !it.promoExpanded) }
    }

    fun confirm() {
        val s = _state.value
        val service = s.selectedService ?: return
        val slot = s.selectedSlot ?: return
        _state.update { it.copy(confirming = true, error = null) }
        viewModelScope.launch {
            runCatching {
                bookings.create(
                    vendorId = vendorId,
                    serviceId = service.id,
                    slotStartIso = slot.startsAt,
                    promoCode = s.promoCodeInput.trim().ifEmpty { null },
                )
            }
                .onSuccess { booking -> _state.update { it.copy(confirming = false, booking = booking) } }
                .onFailure { e -> _state.update { it.copy(confirming = false, error = e.message ?: "Couldn't confirm booking") } }
        }
    }
}
